c/windows.c

```diff
@@ -242,33 +240,46 @@ static IUnknown *s_CreateInstance(CLSID *pCLSID, IID *iid) {
 }
 
 static ptr s_ErrorString(DWORD dwMessageId) {
-    char *lpMsgBuf;
+    char *str = s_ErrorStringImp(dwMessageId, NULL);
+    ptr result = Sstring(str);
+    free(str);
+    return result;
+}
+
+static char *s_ErrorStringImp(DWORD dwMessageId, const char *lpcDefault) {
+    wchar_t *lpMsgBuf;
     DWORD len;
-    ptr result;
+    char *result;
+    char *endstr;
 
-    len = FormatMessage(FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS,
-                        NULL, dwMessageId, MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), (LPTSTR)&lpMsgBuf, 0, NULL);
-    /* If FormatMessage fails, use the error code in hexadecimal. */
+    len = FormatMessageW(FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS,
+                        NULL, dwMessageId, MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), (LPWSTR)&lpMsgBuf, 0, NULL);
+    /* If FormatMessage fails... */
     if (len == 0) {
+        if (lpcDefault) {
+            /* ... use the default string if provided... */
+            return _strdup(lpcDefault);
+        } else {
+            /* ...otherwise, use the error code in hexadecimal. */
 #define HEXERRBUFSIZ ((sizeof(dwMessageId) * 2) + 3)
-        char hexerrbuf[HEXERRBUFSIZ];
-        snprintf(hexerrbuf, HEXERRBUFSIZ, "0x%x", dwMessageId);
-        return Sstring(hexerrbuf);
+            result = malloc(HEXERRBUFSIZ);
+            snprintf(result, HEXERRBUFSIZ, "0x%x", dwMessageId);
+            return result;
 #undef HEXERRBUFSIZ
+        }
     }
+    result = Swide_to_utf8(lpMsgBuf);
+    endstr = result + len;
```

Comment:
len is the number of wide characters in lpMsgBuf, not the number of 8-bit characters in result.

What do you think of pruning the trailing wide characters from lpMsgBuf before converting to UTF-8? Then you can use len in that loop and simply return `Swide_to_utf8(lpMsgBuf)`?

Reply:
I can't believe I missed that! Fixed it like suggested.